frontend_components/src/encord_active_components/frontend/src/components/explorer/Charts.tsx

```diff
@@ -108,23 +112,56 @@ export const ScatteredEmbeddings = ({
   onSelectionChange: (ids: Item2DEmbedding[]) => void;
   onReset: () => void;
 }) => {
+  const binnedItems = useMemo(() => {
+    if (embeddings.length < HEX_BINS)
+      return embeddings.map((emb) => ({ ...emb, size: 1, items: [emb] }));
+
+    const { maxX, minX, maxY, minY } = embeddings.reduce(
+      (maxPoints, { x, y }) => {
+        maxPoints.maxX = Math.max(x, maxPoints.maxX);
+        maxPoints.minX = Math.min(x, maxPoints.minX);
+        maxPoints.maxY = Math.max(y, maxPoints.maxY);
+        maxPoints.minY = Math.min(y, maxPoints.minY);
+        return maxPoints;
+      },
+      { maxX: 0, maxY: 0, minX: 0, minY: 0 }
+    );
+
+    const bins = hexbin<Item2DEmbedding>()
+      .x((d) => d.x)
+      .y((d) => d.y)
+      .radius(((maxX - minX) * (maxY - minY)) / HEX_BINS)(embeddings);
+
+    return bins.map((items) => {
+      const counters = counting(items, ({ label }) => label);
+      const [label] = max(Object.entries(counters), ([_, value]) => value) || [
+        items[0].label,
+        1,
+      ];
+      return { ...items, size: items.length, label, items };
+    });
+  }, [embeddings.length]);
```

Comment:
should probably be just [embeddings] assuming the input is stable. embedding.length only risks failing to trigger in certain rare cases.

Reply:
Make sense 👍